File: src/main/java/eu/yeger/minesweeper/model/Game.java
package eu.yeger.minesweeper.model;

import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;

import java.util.ArrayList;
import java.util.Collection;

public class Game {

    public final int width;
    public final int height;
    public final ObjectProperty<State> state = new SimpleObjectProperty<>(State.IN_PROGRESS);
    public final ArrayList<Cell> cells = new ArrayList<>();

    public Game(final int width,
                final int height) {
        this.width = width;
        this.height = height;
    }

    public Game withCells(final Collection<Cell> cells) {
        cells.forEach(this::withCell);
        return this;
    }

    public Game withCell(final Cell cell) {
        if (!cells.contains(cell)) {
            cells.add(cell);
            cell.doSetGame(this);
        }
        return this;
    }

    public enum State {
        IN_PROGRESS,
        WON,
        LOST
    }
}

File: src/main/kotlin/eu/yeger/minesweeper/model/Game.kt
package eu.yeger.minesweeper.model

import eu.yeger.kofx.property.objectProperty

class Game(
        val width: Int,
        val height: Int
) {
    val stateProperty = objectProperty(State.IN_PROGRESS)
    var state: State by stateProperty

    val cells = ArrayList<Cell>()

    internal fun withCell(cell: Cell) {
        if (cell !in cells) {
            cells.add(cell)
        }
    }

    enum class State {
        IN_PROGRESS,
        WON,
        LOST
    }
}
